File: 1/part2_main.py
def parse_line(line):
    parsed_line = []
    print("LINE ", line.strip())
    index = 0
    while index < len(line):
        if line[index].isdigit():
            parsed_line.append((index, "digit", line[index]))  # (index, is digit)
            index += 1
        else:
            for digit, spelled_digit in enumerate(["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]):
                if line.startswith(spelled_digit, index):
                    parsed_line.append((index, "spelled", spelled_digit,str(digit)))  # (index, is spelled)
                    index += 1
                    break
            else:
                index += 1
    print("PARSED", parsed_line)
    return parsed_line

def process_file(file_path):
    sum = 0
    with open(file_path, 'r') as file:
        for line_number, line in enumerate(file, start=1):
            print(f"Line {line_number}:")
            parsed = parse_line(line)
            first_digit =   parsed[0][2] if parsed[0][1] == 'digit' else parsed[0][3]
            last_digit =   parsed[-1][2] if parsed[-1][1] == 'digit' else parsed[-1][3]

            # Display the results
            print(f"First digit: {first_digit}")
            print(f"Last digit: {last_digit}")
            print(f"inted: ",int(first_digit + last_digit))
            sum = sum + int(first_digit + last_digit)
            print("\n")
    print("sum:", sum)

process_file('input.txt')

File: 1/part2_main.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <iterator>
#include <ranges>

// Used when converting ascii digit to integer
constexpr int ZERO_ASCII_OFFSET = 48;

static auto custom_find_digit_helper(std::string_view line) -> int {
  for (auto it = line.cbegin(); it != line.cend(); it++) {
    if (std::isdigit(*it)) {
      std::cout << "is digit: " << *it << '\n';
      return *it - ZERO_ASCII_OFFSET;
    } else {
      static const std::vector<std::string> spelledOutDigits = {
          "zero", "one", "two",   "three", "four",
          "five", "six", "seven", "eight", "nine"};
      std::string_view from_current_to_end(it, line.cend());
      for (auto index = 0; index < spelledOutDigits.size(); ++index) {
        if (from_current_to_end.starts_with(spelledOutDigits[index])) {
          // std::cout << "string view: " << from_current_to_end << '\n';
          std::cout << "converted: " << index << '\n';
          return index;
        }
      }
    }
  }
  return 0;
}

static auto custom_find_digit_helper_rev(std::string_view line) -> int {
  for (auto it = line.crbegin(); it != line.crend(); it++) {
    if (std::isdigit(*it)) {
      std::cout << "is digit: " << *it << '\n';
      return *it - ZERO_ASCII_OFFSET;
    } else {
      static const std::vector<std::string> spelledOutDigits = {
          "orez", "eno", "owt",   "eerht", "ruof",
          "evif", "xis", "neves", "thgie", "enin"};
      std::string from_current_to_begin(it, line.crend());
      for (auto index = 0; index < spelledOutDigits.size(); ++index) {
        if (from_current_to_begin.starts_with(spelledOutDigits[index])) {
          // std::cout << "string view: " << from_current_to_begin << '\n';
          std::cout << "converted: " << index << '\n';
          return index;
        }
      }
    }
  }
  return 0;
}

int main() {
  namespace rng = std::ranges;
  // Open the file
  std::ifstream inputFile("input.txt");

  if (!inputFile.is_open()) {
    std::cerr << "Error opening the file." << std::endl;
    return 1;
  }

  long int sum = 0;

  // Define a lambda function to process each line
  auto processLine = [&sum](const std::string &line) {
    static int count = 0;
    std::cout << "line: " << line << " coutn: " << ++count << '\n';
    int firstDigit = custom_find_digit_helper(line);
    int lastDigit = custom_find_digit_helper_rev(line);
    // line | std::views::reverse |
    // std::ranges::to<std::vector>()); // use ranges::to<stringview>()  from
    // std23
    sum += 10 * firstDigit + lastDigit;
    std::cout << "sum: " << sum << '\n' << '\n';
  };

  // Read file line by line using ranges and apply the lambda function
  rng::for_each(rng::istream_view<std::string>(inputFile), processLine);

  std::cout << "output: " << sum << std::endl;

  inputFile.close();

  return 0;
}
